src/ControlSystem/Averager.hpp

```diff
@@ -0,0 +1,77 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <array>
+#include <cstddef>
+#include <deque>
+
+#include "DataStructures/DataVector.hpp"
+
+/// \ingroup ControlSystemGroup
+/// A weighted exponential averager of Q and its derivatives
+/// implementing Appendix A in Hemberger et al, https://arxiv.org/abs/1211.6079
+/// The averager is designed to support an arbitrary DerivOrder, however the
+/// private get_derivs method currently only supports DerivOrder = 2.
+/// If an additional DerivOrder is needed, finite differencing needs to be
+/// implemented to specifically handle that order (as it seems like overkill to
+/// generalize the differencing stencil at this time).
+template <size_t DerivOrder>
+class Averager {
+ public:
+  /// `avg_timescale_frac` determines the exponential averaging timescale
+  /// through \f$\tau_\mathrm{avg} = \f$`avg_timescale_frac`\f$\times \tau\f$,
+  /// where \f$\tau\f$ is the damping time.
+  /// `avg_timescale_frac` must be positive.
+  /// `average_0th_deriv_of_q` determines whether the call operator returns an
+  /// averaged or unaveraged quantity for the \f$Q\f$ function values.
+  /// `true` returns an averaged \f$Q\f$. `false` returns the raw \f$Q\f$.
+  /// The derivatives are always averaged, so `average_0th_deriv_of_q` affects
```

Comment:
We could always add a parameter to bypass averaging the derivatives (which would mean that if such a flag is set and average_0th_derv_of_q is false, the entire averager is a no-op), but we've never needed such an option in SpEC.  The derivatives are noisier than the function (because the derivatives are computed by finite-differencing in time) so we generally want to average them.  Averaging the derivatives is the reason we introduced an averager in the first place.   If we really don't want to average the derivatives, then making Averager an abstract base class and having an Averagers::None derived class would be an alternative choice.

Somewhere in the documentation it should be stated that the purpose of the averager is to average out noise in the derivatives of the measured Q parameter, and sometimes average out noise in the Q parameter itself.

Reply:
Right -- I'll try to clarify this in the documentation.